src/Interpreters/Cache/QueryCache.cpp

```diff
@@ -242,8 +242,9 @@ void QueryCache::Writer::finalizeWrite()
         Chunks squashed_chunks;
         size_t rows_remaining_in_squashed = 0; /// how many further rows can the last squashed chunk consume until it reaches max_block_size
 
-        for (const auto & chunk : *query_result)
+        for (auto & chunk : *query_result)
```

Comment:
I don't understand - why don't allow to have sparse columns in the cache?

Reply:
Columns are squashed and we cannot squash sparse and non-sparse columns.

Reply:
Ok.

@rschu1ze, why is there a need for squashing for the query cache? Why don't we save a sequence of blocks as is?

Reply:
Sorry for the late reply.

Squashing is configurable but on by default. The motivation is
1. make entries more compressible when the original query produced many small chunks (because of filtering, aggregation, ...)
3. reading from the cache becomes more "natural" because chunks already have "max_block_size" rows.

The disadvantage is that writing in the cache becomes more expensive. The question is should the cache optimize for insertion or for lookups? With the current approach, the user chooses specific queries which write to the cache (by adding `SETTINGS use_query_cache = 1`). This is different from a model where all or most queries get cached. With selected queries, entries in the cache will likely have a high hit rate, so we should make lookup faster at the cost of more expensive inserts. At least that's the theory 😄 - but in the end that's a heuristics.

Reply:
Good. Let's keep squashing, but also allow cases with many blocks and even with many ports like totals and extremes.